Test/ParallelPrimitives/RadixSort.cpp

```diff
@@ -1,58 +1,121 @@
+#include <Test/OrochiUtils.h>
 #include <Test/ParallelPrimitives/RadixSort.h>
 #include <Test/ParallelPrimitives/RadixSortConfigs.h>
-#include <Test/OrochiUtils.h>
+#include <numeric>
+
+namespace
+{
+
+/// @brief Exclusive scan algorithm on CPU for testing.
+/// It copies the count result from the Device to Host before computation, and then copies the offsets back from Host to Device afterward.
+/// @param countsGpu The count result in GPU memory. Otuput: The offset.
+/// @param offsetsGpu The offsets.
+/// @param nWGsToExecute Number of WGs to execute
+void exclusiveScanCpu( int* countsGpu, int* offsetsGpu, const int nWGsToExecute ) noexcept
+{
+	std::vector<int> counts( Oro::BIN_SIZE * nWGsToExecute );
+	OrochiUtils::copyDtoH( counts.data(), countsGpu, Oro::BIN_SIZE * nWGsToExecute );
+	OrochiUtils::waitForCompletion();
+
+	constexpr auto ENABLE_PRINT{ false };
+
+	if constexpr( ENABLE_PRINT )
+	{
+		for( int j = 0; j < nWGsToExecute; j++ )
+		{
+			for( int i = 0; i < Oro::BIN_SIZE; i++ )
+			{
+				printf( "%d, ", counts[j * Oro::BIN_SIZE + i] );
+			}
+			printf( "\n" );
+		}
+	}
+
+	std::vector<int> offsets( Oro::BIN_SIZE * nWGsToExecute );
+	std::exclusive_scan( std::cbegin( counts ), std::cend( counts ), std::begin( offsets ), 0 );
+
+	OrochiUtils::copyHtoD( offsetsGpu, offsets.data(), Oro::BIN_SIZE * nWGsToExecute );
+	OrochiUtils::waitForCompletion();
+}
+
+} // namespace
 
 namespace Oro
 {
 
 struct RadixSortImpl
 {
-	static
-	void printKernelInfo( oroFunction func ) 
-	{ 
+	static void printKernelInfo( oroFunction func )
+	{
 		int a, b, c;
 		oroFuncGetAttribute( &a, ORO_FUNC_ATTRIBUTE_NUM_REGS, func );
 		oroFuncGetAttribute( &b, ORO_FUNC_ATTRIBUTE_SHARED_SIZE_BYTES, func );
 		oroFuncGetAttribute( &c, ORO_FUNC_ATTRIBUTE_CONST_SIZE_BYTES, func );
 		printf( "vgpr : shared = %d : %d : %d\n", a, b, c );
 	}
 	template<typename T>
-	static
-	void swap( T& a, T& b)
+	static void swap( T& a, T& b )
 	{
 		T t = a;
 		a = b;
 		b = t;
 	}
-
-
 };
 
-#define I RadixSortImpl
+using I = RadixSortImpl;
 
-RadixSort::RadixSort() 
-{ 
+RadixSort::RadixSort()
+{
 	m_nWGsToExecute = 4;
 	m_flags = (Flag)0;
+
+	compileKernels();
+
+	if( selectedScanAlgo == ScanAlgo::SCAN_GPU_PARALLEL )
+	{
+		OrochiUtils::malloc( m_partialSum, m_nWGsToExecute );
+		OrochiUtils::memset( m_partialSum, 0, m_nWGsToExecute * sizeof( int ) );
+	}
 }
 
-RadixSort::~RadixSort()
-{ 
+void RadixSort::compileKernels()
```

Comment:
We need to compile the kernels first and reuse them later in `sort1pass`. Otherwise we will mistakenly include the kernel compilation time in the execution time.